datastream-kafka-connector/src/main/java/com/linkedin/datastream/connectors/kafka/mirrormaker/KafkaMirrorMakerConnectorTask.java

```diff
@@ -382,10 +383,36 @@ protected void maybeCommitOffsets(Consumer<?, ?> consumer, boolean hardCommit) {
   @Override
   protected void postShutdownHook() {
     if (_enablePartitionAssignment) {
-      // The task lock should only be released when it is absolutely safe (we can guarantee that the task cannot
-      // consume any further). The shutdown process must complete and the consumer must be closed.
-      LOG.info("Releasing the lock on datastreamTask: {}", _datastreamTask);
-      _datastreamTask.release();
+      DatastreamRuntimeException exception = null;
+      boolean resetInterrupted = false;
+      for (int numAttempts = 1; numAttempts <= 3; ++numAttempts) {
+        try {
+          // The task lock should only be released when it is absolutely safe (we can guarantee that the task cannot
+          // consume any further). The shutdown process must complete and the consumer must be closed.
+          LOG.info("Releasing the lock on datastreamTask: {}, was thread interrupted: {}, attempt: {}", _datastreamTask,
+              resetInterrupted, numAttempts);
+          _datastreamTask.release();
+          break;
+        } catch (ZkInterruptedException e) {
+          LOG.warn("Releasing the task lock failed for datastreamTask: {}, retrying", _datastreamTask);
+          if (Thread.currentThread().isInterrupted()) {
+            // The interrupted status of the current thread must be reset to allow the task lock to be released
+            resetInterrupted = Thread.interrupted();
+          }
+        } catch (Exception e) {
+          exception = new DatastreamRuntimeException("Failed to perform post shutdown actions", e);
+          break;
+        }
+      }
+
+      if (resetInterrupted) {
+        // Setting the status of the thread back to interrupted
+        Thread.currentThread().interrupt();
+      }
+
+      if (exception != null) {
+        throw exception;
+      }
```

Comment:
Just a suggestion, please feel free to ignore if you don't like it: would it be nicer if wrote this as:
```java
boolean resetInterrupted = false;

try {
    for (int numAttempts = 1; numAttempts <= 3; ++numAttempts) {
        try {
            ...
            _datastreamTask.release();
            break;
        } catch (ZkInterruptedException e) {
            ...
        }
    }
} finally {
    if (resetInterrupted) {
        // Setting the status of the thread back to interrupted
        Thread.currentThread().interrupt();
    }
}
```
Simply put: an outer `try/finally` for resetting interruption status and an inner `try/catch` for retrying lock release. Writing it this way spares you having to add a second `catch` clause with a break statement or an explicit throw.

Reply:
done